File: reference/goathal/maturity.py
"""
Verification Maturity Controller (PROTOCOL layer — device-agnostic). Spec B.

A device class advances CANDIDATE -> PROBATION -> RELAX -> MATURE by holding on-chain
accumulators (V_c, divergence, faults, coverage) over 30-day windows. Sampling relaxes
one step per fully-held window (slow) and SNAPS UP immediately on any breach (fast) — the
asymmetric ratchet. All accumulators are pure functions of published receipts, so anyone
can recompute the window root and fraud-prove an illegal transition (Spec B.4).

Everything here is device-type-agnostic: class_id is an opaque registry string; coverage,
clusters, and ASNs are network/operator concepts, never device types.

Parameters incorporate the accepted fixes: p_floor = 0.15, min slash 15x, cap 20x (F3);
COHORT_MERGE from capability.py (F6) collapses coverage and can force a snap-up.
"""
import hashlib
from dataclasses import dataclass, field
from enum import Enum
from typing import Dict, FrozenSet, List, Optional, Tuple

from .capability import DensitySignal

# ---- constants (F3 fixes) ----
P_FLOOR = 0.15
RELAX_STEPS = (1.0, 0.5, 0.25, P_FLOOR)      # PROBATION starts at 1.0; relax halves to floor
BASE_SLASH = 15.0
SLASH_CAP = 20.0
_EPS = 1e-9

# Stage-1 registration diversity thresholds (nodes, clusters, ASNs, regions)
REG_MIN_NODES, REG_MIN_CLUSTERS, REG_MIN_ASNS, REG_MIN_REGIONS = 50, 25, 10, 5


class Stage(Enum):
    CANDIDATE = 0
    PROBATION = 1
    RELAX = 2
    MATURE = 3


def _rank(s: Stage) -> int:
    return s.value


# ---- receipts & accumulators (Spec B.1) ----
@dataclass(frozen=True)
class Receipt:
    class_id: str
    task_class_id: int
    window: int
    cluster_id: str          # contributing operator cluster (opaque)
    asn: str                 # network ASN (opaque)
    diverged: bool           # result diverged from determinism profile beyond tolerance
    fault: bool              # fault / challenge upheld


@dataclass(frozen=True)
class ClassAccumulator:
    class_id: str
    window: int
    V_c: int
    D_num: int
    F_num: int
    cover_clusters: int
    cover_asns: int

    @property
    def divergence_rate(self) -> float:
        return self.D_num / self.V_c if self.V_c else 0.0

    @property
    def fault_rate_10k(self) -> float:
        return self.F_num / self.V_c * 10_000 if self.V_c else 0.0

    def serialize(self) -> bytes:
        out = bytearray(b"GACC\x01")
        b = self.class_id.encode("utf-8")
        out += len(b).to_bytes(4, "big") + b
        for n in (self.window, self.V_c, self.D_num, self.F_num,
                  self.cover_clusters, self.cover_asns):
            out += int(n).to_bytes(8, "big")
        return bytes(out)

    def root(self) -> bytes:
        return hashlib.sha3_256(self.serialize()).digest()


def _cluster_representative(merge_groups: Optional[List[FrozenSet[str]]]):
    rep: Dict[str, str] = {}
    if merge_groups:
        for grp in merge_groups:
            r = sorted(grp)[0]
            for c in grp:
                rep[c] = r
    return rep


def fold_receipts(receipts: List[Receipt],
                  merge_groups: Optional[List[FrozenSet[str]]] = None
                  ) -> Dict[Tuple[str, int], ClassAccumulator]:
    """Pure fold of published receipts into per-(class, window) accumulators. COHORT_MERGE
    groups (F6) collapse multiple cluster ids into one representative BEFORE coverage is
    counted, so overstated diversity cannot pass the gate."""
    rep = _cluster_representative(merge_groups)
    tmp: Dict[Tuple[str, int], Dict] = {}
    for r in receipts:
        k = (r.class_id, r.window)
        a = tmp.setdefault(k, dict(V=0, D=0, F=0, clusters=set(), asns=set()))
        a["V"] += 1
        a["D"] += 1 if r.diverged else 0
        a["F"] += 1 if r.fault else 0
        a["clusters"].add(rep.get(r.cluster_id, r.cluster_id))
        a["asns"].add(r.asn)
    return {k: ClassAccumulator(k[0], k[1], a["V"], a["D"], a["F"],
                                len(a["clusters"]), len(a["asns"]))
            for k, a in tmp.items()}


def window_root(accumulators: List[ClassAccumulator]) -> bytes:
    """Merkle-ish root over all class accumulators in a window (sorted by class_id)."""
    h = hashlib.sha3_256()
    for a in sorted(accumulators, key=lambda x: x.class_id):
        h.update(a.root())
    return h.digest()


# ---- GATE predicate (Spec B.2) ----
@dataclass(frozen=True)
class GateThresholds:
    v_min: int
    epsilon: float          # max divergence rate
    phi: float              # max fault rate per 10k
    x_clusters: int         # min distinct clusters (coverage)
    x_asns: int             # min distinct ASNs


def gate(acc: ClassAccumulator, th: GateThresholds) -> Tuple[bool, Dict[str, bool]]:
    checks = {
        "volume": acc.V_c >= th.v_min,
        "divergence": acc.divergence_rate < th.epsilon,
        "faults": acc.fault_rate_10k < th.phi,
        "coverage_clusters": acc.cover_clusters >= th.x_clusters,
        "coverage_asns": acc.cover_asns >= th.x_asns,
    }
    return all(checks.values()), checks


# ---- slash sizing coupled to tolerance width (Spec B.5) ----
def slash_multiple(tol_width: float, tol_ref: float,
                   base: float = BASE_SLASH, cap: float = SLASH_CAP,
                   coupling: float = 1.0 / 3.0) -> float:
    """Wider acceptable divergence -> more room to cheat under tolerance -> larger slash.
    `coupling` maps tol_width == tol_ref to the cap (default), rather than the spec's
    implicit coupling=1 which saturates at tol_ref/3 (see clarifications)."""
    if tol_ref <= 0:
        return base
    raw = base * (1.0 + coupling * (tol_width / tol_ref))
    return max(base, min(cap, raw))


def cheat_ev_margin(slash: float, p_effective: float) -> float:
    """cheat_EV < 0 iff slash*p_effective > 1. Returns the margin (safe when > 1)."""
    return slash * p_effective


# ---- state machine (Spec B.3) ----
@dataclass(frozen=True)
class ClassState:
    stage: Stage
    p_class: float
    last_transition_window: int = -1
    slash_mult: float = BASE_SLASH
    pioneer_armed: bool = False


@dataclass(frozen=True)
class Transition:
    class_id: str
    window: int
    from_stage: Stage
    from_p: float
    to_stage: Stage
    to_p: float
    kind: str                 # 'promote' | 'relax' | 'mature' | 'snap' | 'hold'
    gate_ok: bool
    reasons: Tuple[str, ...] = ()


def _snap(stage: Stage, p: float) -> Tuple[Stage, float, str]:
    new_p = min(1.0, p * 2.0)
    new_stage = Stage.PROBATION if new_p >= 1.0 - _EPS else Stage.RELAX
    return new_stage, new_p, "snap"


def evaluate_transition(from_stage: Stage, from_p: float, gate_ok: bool,
                        force_snap: bool) -> Tuple[Stage, float, str]:
    """Pure transition function — shared by the controller and the fraud verifier so they
    cannot disagree. `force_snap` models an immediate intra-window trigger (anomaly burst)."""
    if force_snap and from_stage in (Stage.RELAX, Stage.MATURE) and from_p < 1.0 - _EPS:
        return _snap(from_stage, from_p)
    if from_stage == Stage.CANDIDATE:
        return Stage.CANDIDATE, from_p, "hold"          # leaves CANDIDATE only via registration
    if from_stage == Stage.PROBATION:
        return (Stage.RELAX, 0.5, "promote") if gate_ok else (Stage.PROBATION, 1.0, "hold")
    if from_stage == Stage.RELAX:
        if not gate_ok:
            return _snap(Stage.RELAX, from_p)
        if from_p > P_FLOOR + _EPS:
            return Stage.RELAX, max(P_FLOOR, from_p / 2.0), "relax"
        return Stage.MATURE, P_FLOOR, "mature"
    # MATURE
    if not gate_ok:
        return _snap(Stage.MATURE, from_p)
    return Stage.MATURE, from_p, "hold"


@dataclass(frozen=True)
class RegistrationSet:
    nodes: int
    clusters: int
    asns: int
    regions: int

    def meets_diversity(self) -> bool:
        return (self.nodes >= REG_MIN_NODES and self.clusters >= REG_MIN_CLUSTERS
                and self.asns >= REG_MIN_ASNS and self.regions >= REG_MIN_REGIONS)


class MaturityController:
    def __init__(self, thresholds: GateThresholds, tol_ref: float = 8.0):
        self.th = thresholds
        self.tol_ref = tol_ref
        self.states: Dict[str, ClassState] = {}

    def register_class(self, class_id: str, reg: RegistrationSet, tol_width: float,
                       window: int) -> bool:
        """CANDIDATE -> PROBATION iff the registration set meets diversity thresholds.
        Arms the Pioneer multiplier and sizes slashing to the class tolerance (B.5)."""
        if not reg.meets_diversity():
            self.states[class_id] = ClassState(Stage.CANDIDATE, 1.0, window)
            return False
        self.states[class_id] = ClassState(
            stage=Stage.PROBATION, p_class=1.0, last_transition_window=window,
            slash_mult=slash_multiple(tol_width, self.tol_ref), pioneer_armed=True)
        return True

    def process_window(self, class_id: str, receipts: List[Receipt],
                       merge_groups: Optional[List[FrozenSet[str]]] = None,
                       anomaly_burst: bool = False, window: int = 0
                       ) -> Tuple[Transition, ClassAccumulator]:
        st = self.states[class_id]
        acc = fold_receipts(receipts, merge_groups).get(
            (class_id, window), ClassAccumulator(class_id, window, 0, 0, 0, 0, 0))
        gate_ok, checks = gate(acc, self.th)
        to_stage, to_p, kind = evaluate_transition(st.stage, st.p_class, gate_ok, anomaly_burst)
        reasons = tuple(k for k, v in checks.items() if not v)
        tr = Transition(class_id, window, st.stage, st.p_class, to_stage, to_p, kind,
                        gate_ok, reasons)
        self.states[class_id] = ClassState(to_stage, to_p, window, st.slash_mult, st.pioneer_armed)
        return tr, acc


def cohort_merge_groups(signals: Dict[str, DensitySignal],
                        endpoint_clusters: Dict[str, FrozenSet[str]]
                        ) -> List[FrozenSet[str]]:
    """Translate Item-2 COHORT_MERGE density signals into cluster merge groups the
    accumulator fold uses to collapse overstated diversity."""
    groups = []
    for key, sig in signals.items():
        if sig == DensitySignal.COHORT_MERGE and key in endpoint_clusters:
            groups.append(endpoint_clusters[key])
    return groups


# ---- fraud proof (Spec B.4) ----
@dataclass(frozen=True)
class WindowPosting:
    class_id: str
    window: int
    accumulator_root: bytes
    claimed_from_stage: Stage
    claimed_from_p: float
    claimed_to_stage: Stage
    claimed_to_p: float


@dataclass(frozen=True)
class FraudProof:
    reason: str
    detail: str


def make_posting(tr: Transition, acc: ClassAccumulator) -> WindowPosting:
    return WindowPosting(tr.class_id, tr.window, acc.root(),
                         tr.from_stage, tr.from_p, tr.to_stage, tr.to_p)


def verify_posting(posting: WindowPosting, receipts: List[Receipt],
                   prior: ClassState, th: GateThresholds,
                   merge_groups: Optional[List[FrozenSet[str]]] = None
                   ) -> Optional[FraudProof]:
    """Recompute accumulators from published receipts and detect an illegal posting.
    Fraud iff: (a) the accumulator root does not match; (b) the claimed prior state is
    wrong; (c) the claimed transition is LESS SAFE than the recomputable lower bound —
    i.e. lower sampling (p) or a more advanced stage than the gate justifies. An
    orchestrator may always be MORE conservative (higher p / snap) without being fraudulent."""
    acc = fold_receipts(receipts, merge_groups).get(
        (posting.class_id, posting.window),
        ClassAccumulator(posting.class_id, posting.window, 0, 0, 0, 0, 0))

    if acc.root() != posting.accumulator_root:
        return FraudProof("root_mismatch", "recomputed accumulator root != posted root")

    if (posting.claimed_from_stage, abs(posting.claimed_from_p - prior.p_class) < _EPS) != \
       (prior.stage, True):
        return FraudProof("bad_prior",
                          f"claimed prior {posting.claimed_from_stage.name}/{posting.claimed_from_p} "
                          f"!= known {prior.stage.name}/{prior.p_class}")

    gate_ok, _ = gate(acc, th)
    legal_stage, legal_p, _ = evaluate_transition(prior.stage, prior.p_class, gate_ok, False)

    if posting.claimed_to_p < legal_p - _EPS:
        return FraudProof("undersampling",
                          f"claimed p={posting.claimed_to_p} < legal p={legal_p}")
    if _rank(posting.claimed_to_stage) > _rank(legal_stage):
        return FraudProof("over_advanced",
                          f"claimed stage {posting.claimed_to_stage.name} more advanced than "
                          f"legal {legal_stage.name}")
    return None

File: goatcoin-rs/crates/goat-protocol/src/maturity.rs
//! Verification Maturity Controller (device-agnostic). Spec B with amendments:
//! B-1 (slash coupling = 1/3), B-2 (directional fraud), B-3 (precise snap), B-5 (V_c
//! denominator). Coverage uses the deterministic HLL (R-MAT1) so accumulator roots
//! reproduce across recomputers. class_id/cluster_id/asn are opaque strings.
//!
//! R-MAT2 (recomputable burst snap): the anomaly-burst trigger is DERIVED from the published
//! receipts — each receipt carries an intra-window `sub_window` bucket, the fold tallies
//! anomalous receipts per bucket, and `anomaly_burst()` is a pure integer predicate over that
//! tally. Both the live controller and `verify_posting` recompute it identically, so a posting
//! that withholds a burst-mandated snap is detectable as an illegal (less-safe-than-legal)
//! transition. A declared snap remains possible but can only ADD conservatism.

use std::collections::HashMap;

use sha3::{Digest, Sha3_256};

use crate::capability::DensitySignal;
use crate::hll::Hll;
use crate::types::ClassId;

pub const P_FLOOR: f64 = 0.15;
pub const BASE_SLASH: f64 = 15.0;
pub const SLASH_CAP: f64 = 20.0;
const EPS: f64 = 1e-9;

pub const REG_MIN_NODES: u32 = 50;
pub const REG_MIN_CLUSTERS: u32 = 25;
pub const REG_MIN_ASNS: u32 = 10;
pub const REG_MIN_REGIONS: u32 = 5;

/// R-MAT2: number of intra-window buckets for the recomputable burst tally. A receipt's
/// `sub_window` stamp is folded mod this, so any u32 stamp is safe.
pub const SUB_WINDOWS: u32 = 24;
/// R-MAT2: minimum anomalous receipts in one sub-window before a burst can fire (absolute
/// floor so sparse noise never bursts).
pub const BURST_MIN_EVENTS: u64 = 8;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Stage {
    Candidate = 0,
    Probation = 1,
    Relax = 2,
    Mature = 3,
}

fn rank(s: Stage) -> i32 {
    s as i32
}

#[derive(Clone, Debug)]
pub struct Receipt {
    pub class_id: ClassId,
    pub task_class_id: u32,
    pub window: u64,
    /// Intra-window time bucket (R-MAT2). Folded mod SUB_WINDOWS; lets any recomputer
    /// reconstruct the temporal distribution of anomalies from published receipts.
    pub sub_window: u32,
    pub cluster_id: String,
    pub asn: String,
    pub diverged: bool,
    pub fault: bool,
}

pub struct ClassAccumulator {
    pub class_id: ClassId,
    pub window: u64,
    pub v_c: u64,
    pub d_num: u64,
    pub f_num: u64,
    /// Anomalous receipts (diverged or fault) per sub-window bucket (R-MAT2). Recomputable
    /// from published receipts and bound into `root()`.
    pub sub_events: [u64; SUB_WINDOWS as usize],
    pub cover_clusters_hll: Hll,
    pub cover_asns_hll: Hll,
}

impl ClassAccumulator {
    pub fn empty(class_id: ClassId, window: u64) -> Self {
        Self {
            class_id,
            window,
            v_c: 0,
            d_num: 0,
            f_num: 0,
            sub_events: [0; SUB_WINDOWS as usize],
            cover_clusters_hll: Hll::new(),
            cover_asns_hll: Hll::new(),
        }
    }
    pub fn cover_clusters(&self) -> u64 {
        self.cover_clusters_hll.count()
    }
    pub fn cover_asns(&self) -> u64 {
        self.cover_asns_hll.count()
    }
    pub fn divergence_rate(&self) -> f64 {
        if self.v_c == 0 {
            0.0
        } else {
            self.d_num as f64 / self.v_c as f64
        }
    }
    pub fn fault_rate_10k(&self) -> f64 {
        if self.v_c == 0 {
            0.0
        } else {
            self.f_num as f64 / self.v_c as f64 * 10_000.0
        }
    }
    /// Root over V/D/F, the per-sub-window anomaly tally (R-MAT2), and the HLL REGISTERS
    /// (deterministic) -> reproducible (R-MAT1). The root therefore commits to the TEMPORAL
    /// distribution of anomalies, not just their aggregate counts.
    pub fn root(&self) -> Vec<u8> {
        let mut o = Vec::new();
        o.extend_from_slice(b"GACC\x02");
        let b = self.class_id.as_bytes();
        o.extend_from_slice(&(b.len() as u32).to_be_bytes());
        o.extend_from_slice(b);
        for n in [self.window, self.v_c, self.d_num, self.f_num] {
            o.extend_from_slice(&n.to_be_bytes());
        }
        for n in self.sub_events {
            o.extend_from_slice(&n.to_be_bytes());
        }
        o.extend_from_slice(self.cover_clusters_hll.serialize());
        o.extend_from_slice(self.cover_asns_hll.serialize());
        let mut h = Sha3_256::new();
        h.update(&o);
        h.finalize().to_vec()
    }
}

fn cluster_representative(merge_groups: &[Vec<String>]) -> HashMap<String, String> {
    let mut rep = HashMap::new();
    for grp in merge_groups {
        if let Some(r) = grp.iter().min() {
            for c in grp {
                rep.insert(c.clone(), r.clone());
            }
        }
    }
    rep
}

/// Pure fold of published receipts. COHORT_MERGE groups (F6) collapse cluster ids to one
/// representative BEFORE coverage is counted, so overstated diversity cannot pass the gate.
pub fn fold_receipts(
    receipts: &[Receipt],
    merge_groups: &[Vec<String>],
) -> HashMap<(ClassId, u64), ClassAccumulator> {
    let rep = cluster_representative(merge_groups);
    let mut out: HashMap<(ClassId, u64), ClassAccumulator> = HashMap::new();
    for r in receipts {
        let key = (r.class_id.clone(), r.window);
        let acc = out
            .entry(key)
            .or_insert_with(|| ClassAccumulator::empty(r.class_id.clone(), r.window));
        acc.v_c += 1;
        if r.diverged {
            acc.d_num += 1;
        }
        if r.fault {
            acc.f_num += 1;
        }
        if r.diverged || r.fault {
            acc.sub_events[(r.sub_window % SUB_WINDOWS) as usize] += 1;
        }
        let cluster = rep.get(&r.cluster_id).unwrap_or(&r.cluster_id);
        acc.cover_clusters_hll.add(cluster.as_bytes());
        acc.cover_asns_hll.add(r.asn.as_bytes());
    }
    out
}

// ---- GATE ----
#[derive(Clone, Copy, Debug)]
pub struct GateThresholds {
    pub v_min: u64,
    pub epsilon: f64,
    pub phi: f64,
    pub x_clusters: u64,
    pub x_asns: u64,
}

pub fn gate(acc: &ClassAccumulator, th: &GateThresholds) -> (bool, Vec<(&'static str, bool)>) {
    let checks = vec![
        ("volume", acc.v_c >= th.v_min),
        ("divergence", acc.divergence_rate() < th.epsilon),
        ("faults", acc.fault_rate_10k() < th.phi),
        ("coverage_clusters", acc.cover_clusters() >= th.x_clusters),
        ("coverage_asns", acc.cover_asns() >= th.x_asns),
    ];
    (checks.iter().all(|&(_, v)| v), checks)
}

/// R-MAT2: recomputable anomaly-burst predicate, derived purely from the folded accumulator
/// (never from a declared flag). A burst is a concentration of anomalies in time: one
/// sub-window holds >= BURST_MIN_EVENTS anomalous receipts AND at least half of the window's
/// total. Catches an anomaly episode whose window-wide rates still pass the gate. Integer-only
/// and order-independent, so every recomputer reaches the identical verdict.
pub fn anomaly_burst(acc: &ClassAccumulator) -> bool {
    let total: u64 = acc.sub_events.iter().sum();
    let peak: u64 = acc.sub_events.iter().copied().max().unwrap_or(0);
    peak >= BURST_MIN_EVENTS && 2 * peak >= total
}

// ---- slash sizing (B-1: coupling = 1/3) ----
pub fn slash_multiple(tol_width: f64, tol_ref: f64) -> f64 {
    slash_multiple_cfg(tol_width, tol_ref, BASE_SLASH, SLASH_CAP, 1.0 / 3.0)
}

pub fn slash_multiple_cfg(tol_width: f64, tol_ref: f64, base: f64, cap: f64, coupling: f64) -> f64 {
    if tol_ref <= 0.0 {
        return base;
    }
    let raw = base * (1.0 + coupling * (tol_width / tol_ref));
    raw.clamp(base, cap)
}

pub fn fault_ev_margin(slash: f64, p_effective: f64) -> f64 {
    slash * p_effective
}

// ---- state machine (B-3) ----
#[derive(Clone, Copy, Debug)]
pub struct ClassState {
    pub stage: Stage,
    pub p_class: f64,
    pub last_transition_window: i64,
    pub slash_mult: f64,
    pub pioneer_armed: bool,
}

#[derive(Clone, Debug)]
pub struct Transition {
    pub class_id: ClassId,
    pub window: u64,
    pub from_stage: Stage,
    pub from_p: f64,
    pub to_stage: Stage,
    pub to_p: f64,
    pub kind: &'static str,
    pub gate_ok: bool,
    pub reasons: Vec<&'static str>,
}

fn snap(_stage: Stage, p: f64) -> (Stage, f64, &'static str) {
    let new_p = (p * 2.0).min(1.0);
    let new_stage = if new_p >= 1.0 - EPS {
        Stage::Probation
    } else {
        Stage::Relax
    };
    (new_stage, new_p, "snap")
}

/// Pure transition function — shared by the controller and the fraud verifier.
pub fn evaluate_transition(
    from_stage: Stage,
    from_p: f64,
    gate_ok: bool,
    force_snap: bool,
) -> (Stage, f64, &'static str) {
    if force_snap && matches!(from_stage, Stage::Relax | Stage::Mature) && from_p < 1.0 - EPS {
        return snap(from_stage, from_p);
    }
    match from_stage {
        Stage::Candidate => (Stage::Candidate, from_p, "hold"),
        Stage::Probation => {
            if gate_ok {
                (Stage::Relax, 0.5, "promote")
            } else {
                (Stage::Probation, 1.0, "hold")
            }
        }
        Stage::Relax => {
            if !gate_ok {
                snap(Stage::Relax, from_p)
            } else if from_p > P_FLOOR + EPS {
                (Stage::Relax, (from_p / 2.0).max(P_FLOOR), "relax")
            } else {
                (Stage::Mature, P_FLOOR, "mature")
            }
        }
        Stage::Mature => {
            if !gate_ok {
                snap(Stage::Mature, from_p)
            } else {
                (Stage::Mature, from_p, "hold")
            }
        }
    }
}

#[derive(Clone, Copy, Debug)]
pub struct RegistrationSet {
    pub nodes: u32,
    pub clusters: u32,
    pub asns: u32,
    pub regions: u32,
}

impl RegistrationSet {
    /// Mainnet default thresholds (50/25/10/5).
    pub fn mainnet_min() -> Self {
        Self {
            nodes: REG_MIN_NODES,
            clusters: REG_MIN_CLUSTERS,
            asns: REG_MIN_ASNS,
            regions: REG_MIN_REGIONS,
        }
    }
    pub fn meets(&self, min: &RegistrationSet) -> bool {
        self.nodes >= min.nodes
            && self.clusters >= min.clusters
            && self.asns >= min.asns
            && self.regions >= min.regions
    }
    pub fn meets_diversity(&self) -> bool {
        self.meets(&RegistrationSet::mainnet_min())
    }
}

pub struct MaturityController {
    pub th: GateThresholds,
    pub tol_ref: f64,
    pub reg_min: RegistrationSet,
    pub states: HashMap<ClassId, ClassState>,
}

impl MaturityController {
    pub fn new(th: GateThresholds, tol_ref: f64) -> Self {
        Self {
            th,
            tol_ref,
            reg_min: RegistrationSet::mainnet_min(),
            states: HashMap::new(),
        }
    }

    /// Controller with scaled registration thresholds (e.g. for a small testnet).
    pub fn with_reg_min(th: GateThresholds, tol_ref: f64, reg_min: RegistrationSet) -> Self {
        Self {
            th,
            tol_ref,
            reg_min,
            states: HashMap::new(),
        }
    }

    pub fn register_class(
        &mut self,
        class_id: &str,
        reg: RegistrationSet,
        tol_width: f64,
        window: i64,
    ) -> bool {
        if !reg.meets(&self.reg_min) {
            self.states.insert(
                class_id.to_string(),
                ClassState {
                    stage: Stage::Candidate,
                    p_class: 1.0,
                    last_transition_window: window,
                    slash_mult: BASE_SLASH,
                    pioneer_armed: false,
                },
            );
            return false;
        }
        self.states.insert(
            class_id.to_string(),
            ClassState {
                stage: Stage::Probation,
                p_class: 1.0,
                last_transition_window: window,
                slash_mult: slash_multiple(tol_width, self.tol_ref),
                pioneer_armed: true,
            },
        );
        true
    }

    /// Process a window's receipts. The burst snap is RECOMPUTED from the receipts (R-MAT2);
    /// `declared_snap` can add a conservative extra snap but can never suppress the
    /// recomputable one.
    pub fn process_window(
        &mut self,
        class_id: &str,
        receipts: &[Receipt],
        merge_groups: &[Vec<String>],
        declared_snap: bool,
        window: u64,
    ) -> (Transition, Vec<u8>) {
        let st = *self.states.get(class_id).expect("class registered");
        let folded = fold_receipts(receipts, merge_groups);
        let empty;
        let acc = match folded.get(&(class_id.to_string(), window)) {
            Some(a) => a,
            None => {
                empty = ClassAccumulator::empty(class_id.to_string(), window);
                &empty
            }
        };
        let root = acc.root();
        let (gate_ok, checks) = gate(acc, &self.th);
        let force_snap = declared_snap || anomaly_burst(acc);
        let (to_stage, to_p, kind) = evaluate_transition(st.stage, st.p_class, gate_ok, force_snap);
        let reasons: Vec<&'static str> = checks
            .iter()
            .filter(|&&(_, v)| !v)
            .map(|&(k, _)| k)
            .collect();
        let tr = Transition {
            class_id: class_id.to_string(),
            window,
            from_stage: st.stage,
            from_p: st.p_class,
            to_stage,
            to_p,
            kind,
            gate_ok,
            reasons,
        };
        self.states.insert(
            class_id.to_string(),
            ClassState {
                stage: to_stage,
                p_class: to_p,
                last_transition_window: window as i64,
                slash_mult: st.slash_mult,
                pioneer_armed: st.pioneer_armed,
            },
        );
        (tr, root)
    }
}

/// Translate Item-2 COHORT_MERGE density signals into cluster merge groups.
pub fn cohort_merge_groups(
    signals: &HashMap<ClassId, DensitySignal>,
    endpoint_clusters: &HashMap<ClassId, Vec<String>>,
) -> Vec<Vec<String>> {
    let mut groups = Vec::new();
    for (key, sig) in signals {
        if *sig == DensitySignal::CohortMerge {
            if let Some(clusters) = endpoint_clusters.get(key) {
                groups.push(clusters.clone());
            }
        }
    }
    groups
}

// ---- fraud proof (B-2: directional) ----
#[derive(Clone, Debug)]
pub struct WindowPosting {
    pub class_id: ClassId,
    pub window: u64,
    pub accumulator_root: Vec<u8>,
    pub claimed_from_stage: Stage,
    pub claimed_from_p: f64,
    pub claimed_to_stage: Stage,
    pub claimed_to_p: f64,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct FraudProof {
    pub reason: &'static str,
    pub detail: String,
}

pub fn make_posting(tr: &Transition, root: Vec<u8>) -> WindowPosting {
    WindowPosting {
        class_id: tr.class_id.clone(),
        window: tr.window,
        accumulator_root: root,
        claimed_from_stage: tr.from_stage,
        claimed_from_p: tr.from_p,
        claimed_to_stage: tr.to_stage,
        claimed_to_p: tr.to_p,
    }
}

/// Recompute accumulators from published receipts and detect an illegal posting.
/// Fraud iff: root mismatch, wrong claimed prior, or a LESS SAFE claim than the
/// recomputable lower bound (lower sampling p, or more-advanced stage). Conservatism
/// (higher p / snap) is never fraudulent.
///
/// R-MAT2: the anomaly-burst snap is part of the recomputable lower bound. The verifier
/// derives the burst from the receipts (`anomaly_burst`) exactly as the controller does, so
/// a posting that withholds a burst-mandated snap is illegal and is reported with the
/// precise reason `withheld_burst_snap` when the claim matches the legal NON-burst
/// transition (i.e. the violation is explained entirely by the withheld snap).
pub fn verify_posting(
    posting: &WindowPosting,
    receipts: &[Receipt],
    prior: &ClassState,
    th: &GateThresholds,
    merge_groups: &[Vec<String>],
) -> Option<FraudProof> {
    let folded = fold_receipts(receipts, merge_groups);
    let empty;
    let acc = match folded.get(&(posting.class_id.clone(), posting.window)) {
        Some(a) => a,
        None => {
            empty = ClassAccumulator::empty(posting.class_id.clone(), posting.window);
            &empty
        }
    };
    let root = acc.root();
    let (gate_ok, _) = gate(acc, th);
    let burst = anomaly_burst(acc); // recomputed from receipts, never declared

    if root != posting.accumulator_root {
        return Some(FraudProof {
            reason: "root_mismatch",
            detail: "recomputed root != posted root".into(),
        });
    }
    if posting.claimed_from_stage != prior.stage
        || (posting.claimed_from_p - prior.p_class).abs() >= EPS
    {
        return Some(FraudProof {
            reason: "bad_prior",
            detail: format!(
                "claimed prior {:?}/{} != known {:?}/{}",
                posting.claimed_from_stage, posting.claimed_from_p, prior.stage, prior.p_class
            ),
        });
    }
    let (legal_stage, legal_p, _) = evaluate_transition(prior.stage, prior.p_class, gate_ok, burst);
    // A violation explained entirely by the recomputed burst (the claim IS the legal
    // non-burst transition or safer) gets the precise reason.
    let (base_stage, base_p, _) = evaluate_transition(prior.stage, prior.p_class, gate_ok, false);
    let withheld_snap =
        |cs: Stage, cp: f64| burst && cp >= base_p - EPS && rank(cs) <= rank(base_stage);
    if posting.claimed_to_p < legal_p - EPS {
        let reason = if withheld_snap(posting.claimed_to_stage, posting.claimed_to_p) {
            "withheld_burst_snap"
        } else {
            "undersampling"
        };
        return Some(FraudProof {
            reason,
            detail: format!(
                "claimed p={} < legal p={} (recomputed burst={})",
                posting.claimed_to_p, legal_p, burst
            ),
        });
    }
    if rank(posting.claimed_to_stage) > rank(legal_stage) {
        let reason = if withheld_snap(posting.claimed_to_stage, posting.claimed_to_p) {
            "withheld_burst_snap"
        } else {
            "over_advanced"
        };
        return Some(FraudProof {
            reason,
            detail: format!(
                "claimed {:?} more advanced than legal {:?} (recomputed burst={})",
                posting.claimed_to_stage, legal_stage, burst
            ),
        });
    }
    None
}
